parseAPI/src/IA_x86.C

```diff
@@ -91,68 +94,81 @@ bool IA_x86::isFrameSetupInsn(Instruction i) const
     return false;
 }
 
-class nopVisitor : public InstructionAPI::Visitor
-{
-    public:
-        nopVisitor() : foundReg(false), foundImm(false), foundBin(false), isNop(true) {}
-        virtual ~nopVisitor() {}
 
-        bool foundReg;
-        bool foundImm;
-        bool foundBin;
-        bool isNop;
+class leaSimplifyVisitor : public InstructionAPI::Visitor {
+        using ExprNodeValues = boost::variant2::variant<Expression*, RegisterAST*, uint64_t>;
+        using ExprStack = std::stack<ExprNodeValues>;
+        ExprStack exprStack;
 
-        virtual void visit(BinaryFunction*)
-        {
-            if (foundBin) isNop = false;
-            if (!foundImm) isNop = false;
-            if (!foundReg) isNop = false;
-            foundBin = true;
+    public:
+        void visit(BinaryFunction *bf) override {
+            auto op1 = exprStack.top();
+            exprStack.pop();
+            auto op2 = exprStack.top();
+            exprStack.pop();
+
+            auto value = boost::variant2::get_if<uint64_t>(&op1);
+            if (!value)  {
+                std::swap(op1, op2);  // op1 not an immediate, swap and try what was in op2
+                value = boost::variant2::get_if<uint64_t>(&op1);
+            }
+            if (value && ((*value == 0 && bf->isAdd()) || (*value == 1 && bf->isMultiply())))  {
+                exprStack.push(op2);  // additive or multiplicative identity, simplify to op2
+            }  else  {
+                exprStack.push(bf);   // no simplification, use BinaryFunction
+            }
         }
-        virtual void visit(Immediate *imm)
-        {
-            if (imm != 0) isNop = false;
-            foundImm = true;
+        void visit(Immediate *imm) override {
+            auto const value = imm->eval().convert<uint64_t>();
```

Comment:
It won't make any difference to the calculation in `visit(BinaryFunction*)`, but displacements can be signed values. The scalar field of a SIB byte is only positive.

Reply:
Changed.